# Test 4.7 Deriving New Variables

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

This script contains the following points:
1. Importing of combined orders_products data
2. Creating a subset population
3. Defining new user-defined fields for price_range, busiest_days, busiest_period_of_day

In [2]:
path=r'C:\Users\dell\Desktop\Career Foundry\0423_Instacart Analysis'

In [3]:
# Import pickle for order_products_combined data set
df_orders_prods_merged=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_prods_merged.pkl'))

In [4]:
# Create a subset of df_orders_prods_merged
df_opm_subset=df_orders_prods_merged[:1000000]

In [5]:
df_opm_subset.shape

(1000000, 13)

In [6]:
# Define a user-defined price label
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
# Apply price_label function

df_opm_subset['price_range']=df_opm_subset.apply(price_label, axis=1)

C:\Users\dell\AppData\Local\Temp\ipykernel_12996\3462624236.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_opm_subset['price_range']=df_opm_subset.apply(price_label, axis=1)


In [8]:
df_opm_subset['price_range'].value_counts(dropna=False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [9]:
df_opm_subset['prices']. max()

14.8

In [10]:
df_opm_subset.loc[df_opm_subset['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\dell\AppData\Local\Temp\ipykernel_12996\1852137448.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_opm_subset.loc[df_opm_subset['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [11]:
df_opm_subset.loc[(df_opm_subset['prices'] <= 15) & (df_opm_subset['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [12]:
df_opm_subset.loc[df_opm_subset['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
df_opm_subset['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [14]:
df_orders_prods_merged.loc[df_orders_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [15]:
df_orders_prods_merged.loc[(df_orders_prods_merged['prices'] <= 15) & (df_orders_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [16]:
df_orders_prods_merged.loc[df_orders_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
df_orders_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

In [18]:
df_orders_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   #ofdays_since_last_order  product_id  add_to_cart_order  reordered  \
0                       0.0         196                  1          0   
1                      15.0         196                  1          1   
2                      21.0         196                  1          1   
3                      29.0         196                  1          1   
4                      28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  
0         Soda        77              7     9.0  Mid-range product  
1         Soda        77              7     9.0  Mid-range product  
2         Soda        77              7     9.0  Mid-range product  
3         Soda        77              7     9.0  Mid-range product  
4         Soda        77              7     9.0  Mid-range product

In [19]:
df_orders_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [20]:
# Defining busy day of the week loop
results = []

for value in df_orders_prods_merged["orders_day_of_week"]:
  if value == 0:
    results.append("Most busy")
  elif value == 4:
    results.append("Least busy")
  else:
    results.append("Average busy")

In [21]:
df_orders_prods_merged['busiest_day']=results

In [22]:
df_orders_prods_merged['busiest_day'].value_counts(dropna=False)

Average busy    22416875
Most busy        6204182
Least busy       3783802
Name: busiest_day, dtype: int64

In [23]:
# Defining busiest days(plural) of the week loop
new_result = []

for value in df_orders_prods_merged["orders_day_of_week"]:
  if (value == 0 or value == 1):
    new_result.append("Busiest Days")
  elif (value == 4 or value ==3):
    new_result.append("Slowest Days")
  else:
    new_result.append("Average Days")

In [24]:
df_orders_prods_merged['busiest_days']=new_result

In [25]:
df_orders_prods_merged['busiest_days'].value_counts(dropna=False)

Average Days    12916111
Busiest Days    11864412
Slowest Days     7624336
Name: busiest_days, dtype: int64

# Average Days decreased from ~22 mil down to ~12 mil. Sum of values 0 and 1 equals 11,864,412 and sum of values 4 and 3 equals 7,624,336, proving the updated loop worked. 

In [26]:
df_orders_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [27]:
# Defining busiest hours(plural) of the day loop
hour_result = []

for value in df_orders_prods_merged["order_hour_of_day"]:
  if (value == 10 or value == 11):
    hour_result.append("Most orders")
  elif (value == 4 or value ==3):
    hour_result.append("Fewest orders")
  else:
    hour_result.append("Average orders")

In [28]:
df_orders_prods_merged['busiest_period_of_day']=hour_result

In [29]:
df_orders_prods_merged['busiest_period_of_day'].value_counts(dropna=False)

Average orders    26802436
Most orders        5497900
Fewest orders       104523
Name: busiest_period_of_day, dtype: int64

In [30]:
df_orders_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   #ofdays_since_last_order  product_id  add_to_cart_order  reordered  \
0                       0.0         196                  1          0   
1                      15.0         196                  1          1   
2                      21.0         196                  1          1   
3                      29.0         196                  1          1   
4                      28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

    busiest_day  busiest_days busiest_period_of_day  
0  Average busy  Average Days        Average orders  
1  Average busy  Slowest Days        Average orders  
2  Average busy  Slowest Days        Average orders  
3    Least busy  Slowest Days        Average orders  
4    Least busy  Slowest Days        Average orders

In [31]:
# Removing orginal busiest_day column from dataframe
df_ord_prod_newvar=df_orders_prods_merged.drop('busiest_day',1)

C:\Users\dell\AppData\Local\Temp\ipykernel_12996\2248573567.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df_ord_prod_newvar=df_orders_prods_merged.drop('busiest_day',1)


In [32]:
df_ord_prod_newvar.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   #ofdays_since_last_order  product_id  add_to_cart_order  reordered  \
0                       0.0         196                  1          0   
1                      15.0         196                  1          1   
2                      21.0         196                  1          1   
3                      29.0         196                  1          1   
4                      28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

   busiest_days busiest_period_of_day  
0  Average Days        Average orders  
1  Slowest Days        Average orders  
2  Slowest Days        Average orders  
3  Slowest Days        Average orders  
4  Slowest Days        Average orders

In [33]:
# Export data to pkl

df_ord_prod_newvar.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ord_prod_newvar.pkl'))